## Bitvisa opperationer i Python
Det finns många tillfällen där vi vill arbeta med de enskilda bitarna i en variabel. För att kunna göra det så har vi till vårt förfogande de bitvisa operatorerna. Dessa är:

| Operator  | Namn  | Funktion  |
|---|---|---|
| & | AND |Logiskt OCH   |
| \| | OR  |Logiskt ELLER   |
| ^ | XOR  |Exlusivt ELLER   |
| ~ | NOT  |Invertera alla bitar   |
| << | Vänster skift  |Flyttar alla bitar n steg åt vänster, fyller på med nollor   |
| >> | Höger skift  |Flyttar alla bitar n steg åt höger, fyller på med nollor   |


Innan vi kan förstå hur dessa operatorer fungerar måste vi bekanta oss med två andra saker, det binära talsystemet och hur en byte ser ut och fungerar.

### Det binära talsystemet
Det binära talsystemet har talbasen 2, istället för basen 10 som vi vanligtvis använder. När vi räknar med våra 10 siffror blir resultatet fölande:

0
1
2
3
4
5
6
7
8
9

Nu har vi slut på siffror, så vad gör vi? Jo vi börjar om på 0, men sätter en etta framför:

10
11
osv

När vi arbetar med det binära talsystemet gör vi samma sak, med den enda skillnaden att vi har två siffror istället för 10, 0 och 1

0
1

Nu har vi slut på siffror så vi börjar om och sätter en etta före

10
11

Samma sak upprepas nu

100
101
110
111

och så håller vi på så.

### En byte
En bit är den minsta enheten vi kan jobba med i en dator och den representerar antingen en nolla eller en etta.

Bitarna klumpas ihop i grupper om åtta, och kallas då en byte. Åtta bitar kan nu tillsammans bilda ett större tal.

För att öka läsbarheten brukar man skriva en byte i två grupperingar, med fyra bitar i varje. Detta är alltså inget datorn gör internt utan bara något vi gör för att det skall bli enklare att se vilken bit som är vilken. En byte kan då representeras så här:

1011 0110

Om vi tittar på en byte så kan vi tänka oss att varje position representerar ett decimalt värde. En etta på motsvarande plats betyder att detta värde är en del av det tal som byten representerar och en nolla att det inte är det.

Då ser samma byte som ovan ut så här:

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 

Vi ser att vi har ettor vid 128, 32, 16, 4 och 2. Denna byte representerar alltså talet 128 + 32 + 16 + 4 + 2 som är 182.

Vi kan också konstatera att alla udda tal måste ha en etta längst till höger (i rutan som represeterar värdet 1) eftersom det är det enda udda värdet som vi har. Detta kommer väl tillhands i nästa avsnitt.

### Logiskt OCH
Vi kan jämföra två bytes med de logiska operatorerna AND (&) och OR (\|).

Om vi tittar på AND först så jämför den två bytes bit för bit. Resutatet kommer att bli en ny byte, som har en etta på alla platser där de två ursprungliga bytsen hadde en etta, annars kommer den att ha en nolla på den platsen. Vi tar ett exempel för att förtydliga.

Vi börjar med samma byte som vi hade tidigare:

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 

Sen tar vi en annan byte som kanske ser ut så här:

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 0 | 0 | 1 | 1 | 1 | 0 | 0 | 0 | 

Om vi nu utför ett bitvis OCH på dessa två så kommer vi att få en ny byte som har en etta på alla platser där de båda har en etta. Om bara en eller ingen av de två har en etta på den här platsen blir resutatet en nolla.

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 
| 0 | 0 | 1 | 1 | 1 | 0 | 0 | 0 | 
| F | F | S | S | F | F | F | F | 

F och S i tabellen ovan indikerar om resultatet blir sant eller falsk för respektive bit-par.

Resultatet blir nu en byte som har ett på varje position som är markerad med ett S, alltså:

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 0 | 0 | 1 | 1 | 0 | 0 | 0 | 0 | 

Vi skulle nu kunna använda detta för att kontrollera om en viss bit är satt (att den är satt betyder att den har värdet 1) i en byte.

Om vi tänker oss följande. Vi har en variabel som har värdet 13 och vill kolla om det är ett udda eller jämnt tal, då vet vi att om det är ett udda tal så måste det finnas en etta i den högraste biten. Annars är det ett jämnt tal. 

13 binärt blir

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 0 | 0 | 0 | 0 | 1 | 1 | 0 | 1 | 

Om vi nu vill kolla om den bit som representerar 1 är satt eller inte så kan vi göra ett binärt AND med en byte där vi vet att den är det, förslagsvis en byte som representerar värdet 1.

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 0 | 0 | 0 | 0 | 1 | 1 | 0 | 1 | 
| 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 
| F | F | F | F | F | F | F | S | 

Blir den resulterande byten 1 så är det ett udda tal och blir den 0 så är det ett jämnt tal.

I detta fallet blir det ett, 13 är alltså ett udda tal. 

Låt oss kolla detta i kod.

In [23]:
value = 13

result = value & 1

print(result)

1


Provar vi nu samma sak med ett jämnt tal så ser vi att resultatet blir 0.

In [24]:
value = 12

result = value & 1

print(result)

0


Vi kan nu använda oss av det faktum att 0 betraktas som falskt i en vanlig if-sats och att alla andra värden är sant.

In [25]:
value = 13

if value & 1:
    print('This is an odd value')
else:
    print('This value is even')

This is an odd value


### Logiskt ELLER
Logiskt eller fungerar precis på samma vis som logiskt OCH, med en skillnad. Resultatet blir 1 och en eller båda bitarna som jämförs är 1. Så ett ELLER på de två bytes vi såg tidigare skulle resultera i följande:

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|---|---|---|---|---|---|---|---|
| 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 
| 0 | 0 | 1 | 1 | 1 | 0 | 0 | 0 | 
| S | F | S | S | S| S | S | F | 

Vilket då ger oss följande byte:

1011 1110

Vi kan alltså använda denna operator för att kolla om en eller båda bitarna är satta.

I kod skulle det se ut så här:

In [26]:
value = 182
result =  value | 56
print(result)

190


Vi ser att vi får 190 vilket är 1011 1110 binärt.

### Exlusivt ELLER
Exlusivt eller, även kallat XOR, fungerar som OR och AND, men för att resultatet skall bli 1 för en bit så måste de bitar som jämförs vara en av varje. Så om vi har 1 i den ena byten på en plats och 0 på samma plats i vår andra byte så blir resultatet 1. I alla andra fall blir resultatet 0.

Vad kan man då använda XOR til? Faktum är att den är väldigt användbar eftersom den har vissa unika egenskaper. Låt oss kolla ett exempel.

Låt säga att vi vill representera bokstaven j i en byte. I ASCII-tabellen har lilla j värdet 106. 
Binärt blir 106 0110 1010.

Så det är vårt ena värde. Sen kan vi ta en annan bokstav, säg lilla u som i ASCII-tabellen har värdet 117.

Binärt blir det 0111 0101.

Låt oss göra en XOR-operation på dessa två bytes.

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 | | |
|---|---|---|---|---|---|---|---|---|---|
| 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | -> | j | 
| 0 | 1 | 1 | 1 | 0 | 1 | 0 | 1 | -> | u |
| F | F | F | S | S| S | S | S | 

Vi ser nu att resultatet blir 0001 1111.

Men vad händer om vi tar detta värde och gör ett XOR med den byte som represeterar vårt u?

Jo vi får detta:
| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 | | |
|---|---|---|---|---|---|---|---|---|---|
| 0 | 0 | 0 | 1 | 1 | 1 | 1 | 1 |  |  | 
| 0 | 1 | 1 | 1 | 0 | 1 | 0 | 1 | -> | u |
| F | S | S | F | S| F | S | F | ||

Resultatet blir alltså 0110 1010, eller 106 som råkade representera lilla j. Så vi fick tillbaka vårt ursprungliga j igen. 

Detta gör att vi kan använda detta för en enkel form av kryptering. I vårt lilla exempel skulle j vara det vi vill kryptera och u skulle vara vårt lösenord. 0001 1111, eller 31 decimalt blir vårt krypterade meddelande.

Nu är XOR i sig själv inte en jättesäker krypteringsmetod, men den används som en del av moderna avancerade krypteringsalgoritmer.

Vill vi se detta i action så blir det så här. Vi använder här funktionen ord() för att få fram ett teckens ASCII-nummer och functionen chr() för att få fram tecknet från ett ASCII-nummer.

In [27]:
clear_text = 'j'
password = 'u'

# Encrypt with XOR
result = ord(clear_text) ^ ord(password)

print(f'The encrypted value is {result}')

# Decrypt with XOR

decrypted = result ^ ord(password)

print(f'The decrypted message is {chr(decrypted)}')

The encrypted value is 31
The decrypted message is j


Ett enkelt krypteringsprogram i Python som använder XOR skulle kunna se ut så här:

In [28]:
from getpass import getpass 

clear_text = input("Enter text to be encrypted: ")
password = getpass()

password = list(password)

# Encrypt
encrypted = []
for i, character in enumerate(clear_text):
    encrypted.append(ord(character) ^ ord(password[i % len(password)]))

print(f'Encrypted data is {encrypted}')

password = getpass()

password = list(password)

# Decrypt
decrypted_message = []
for i, cipher_value in enumerate(encrypted):
    decrypted_message.append(chr(cipher_value ^ ord(password[i % len(password)])))
decrypted_message = ''.join(decrypted_message)
print(decrypted_message)

Encrypted data is [39, 13, 10, 1, 69, 29, 0, 69, 2, 82, 22, 17, 16, 23, 6, 6, 69, 25, 22, 22, 16, 19, 2, 17]
This is a secret message


Notera att om du vill köra detta programmet så funkar det bäst att kopiera koden till en vanlig editor och köra det där.

### Negation
Negation betyder helt enkelt att alla bitar som är 1 blir 0 och vice versa. Enklast är kanske att bara se detta i kod.

In [29]:
value = 123
new_value = ~value

print(new_value)

value = ~new_value
print(value)



-124
123


Vi ser här att det negerade värdet av 123 blir -124. Anledningen till att det blir ett negativt tal handlar om något som heter tvåkomplementsmetoden och som är det som används för att representera negativa tal. Vill du lära dig mer om den så kan du antingen läsa om den i din väns bok, eller läsa denna korta beskrivning.

#### Tvåkoplementmetoden eller tvåkoplementformen
För att kunna representera negativa tal så låter vi den vänstra biten, kallas också den mest signifikanta biten, ha värdet 1 för ett negativt tal och 0 för ett positivt. Om vi använder denna metoden (viklet datorn gör) för att representera tal så ser positionernas värden i en byte lite annorlunda ut.

| -128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 | 
|---|---|---|---|---|---|---|---|
| 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |

Vi ser nu att den mest signifikanta biten inte längre representerar 128, utan -128.

Om vi tittar på talet vi negerade i exemplet tidigare så får vi följande:

Vi började med talet 123, vilket är 0111 1011 binärt.
Vi negerade detta och fick då det binära talet 1000 0100.

När vi skrev ut det så fick vi värdet -124. Hur gick detta till. 

Vi ser att vi har en etta på platsen för -128. Detta är alltså ett negativt tal. Vi har även en etta på platsen för 4. Adderar vi ihop dessa (-128 + 4) så får vi -124. Vi ser också att om vi använder oss av denna metod så ändras också omfånget för vilka tal vi kan representera. När vi bara körde med positiva tal var det största talet vi kunde ha:

1111 1111, vilket är 255

Om nu den mest signifikanta biten skall användas för att indikera positiva eller negativa tal så återstår 7 bitar och det största talet vi kan ha är 127.

Det minsta talet vi kan representera blir då -128, vilket binärt blir 1000 000 eftersom -128 minus inget är -128.

Det intressanta är att om vi vill representera -1 så blir det

1111 1111

-128 + 64 + 32 + 16 + 8 + 4 + 2 + 1

vilket blir -1.

### Bitvis skiftning
Vi kan också flytta alla nollor och ettor ett visst antal steg antingen åt höger eller åt vänster. Detta gör vi med skiftoperatorerna << och >>. Den första skiftar bitarna åt vänster, den andra åt höger. När vi flyttar alla bitar ett steg kommer en bit att "falla ut" och på den andra sidan kommer en nolla att stoppas in. Om vi har denna byte 

| -128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 | 
|---|---|---|---|---|---|---|---|
| 0 | 1 | 0 | 0 | 0 | 1 | 0 | 1 |

och ckiftar den ett steg åt höger så kommer vi att få 


| -128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 | 
|---|---|---|---|---|---|---|---|
| 0 | 0 | 1 | 0 | 0 | 0 | 1 | 0 |

Vi ser att den etta som stod längst till höger faller bort, att alla andra bitar flyttat sig ett steg åt höger och att vi har stoppat in en nolla längst till vänster.

Syntaxen för att skifta är

value = value >> 1

Detta flyttar alla bitar ett steg åt höger och lagrar det nya värdet i value igen.

value = value >> 3

gör samma sak men flyttar bitarna tre steg.

Låt oss kolla detta i kod

In [30]:
value = 64
value = value >> 1
print(value)

32


Vi ser att biten som först stod på platsen som representerar 64 nu flyttat ett steg och nu står på platsen som reprsenterar 32.

Vad kan vi då använda detta till? Ett exempel är om vi vill kolla hur ett decimalt tal representeras binärt. Vi kan då skifta bitarna steg för steg och kolla den högraste biten (som förövrigt brukar kallas den minst signifikanta biten) mot 1. Funkar fint, förutom att det binära talet vi får ut blir baklänges.

Vi kan också lägga till att vi kan använda den mer kompakta formen

value >>= 1

istället för 

value = value >> 1

In [31]:
value = 89  # 89 binary is 0101 1001

for _ in range(8):
    if value & 1:
        print('1', end='')
    else:
        print('0', end='')
    value >>= 1

10011010

Vi ser att vi fått rätt binärt tal, men att det är baklänges. Vi kan ju lösa detta genom att lagra resultatet i en sträng och skriva ut den baklänges när vi är klara.

In [32]:
value = 89  # 89 binary is 0101 1001

bin_str = ''
for _ in range(8):
    if value & 1:
        bin_str += '1'
    else:
        bin_str += '0'
    value >>= 1

print(bin_str[::-1])

01011001
